autofit/graphical/messages/abstract.py

```diff
@@ -1,256 +1,389 @@
+        that match the sufficient statistics
+        """
+        # if weights aren't passed then equally weight all samples
+
+        # Numerically stable weighting for very small/large weights
+        log_w_max = np.max(log_weights, axis=0, keepdims=True)
+        w = np.exp(log_weights - log_w_max)
+        norm = w.mean(0)
+        log_norm = np.log(norm) + log_w_max[0]
+
+        tx = cls.to_canonical_form(samples)
+        w /= norm
+        suff_stats = (tx * w[None, ...]).mean(1)
+
+        assert np.isfinite(suff_stats).all()
+        return cls.from_sufficient_statistics(suff_stats, log_norm=log_norm)
+
+    @classmethod
+    def from_mode(cls, mode, covariance):
+        pass
+
+    def log_normalisation(self, *dists: Union["AbstractMessage", float]) -> np.ndarray:
+        """
+        Calculates the log of the integral of the product of a
+        set of distributions
+
+        NOTE: ignores log normalisation
+        """
+        # Remove floats from messages passed
+        dists = [
+            dist for dist in self._iter_dists(dists) 
+            if isinstance(dist, AbstractMessage)]
+
+        # Calculate log product of message normalisation
+        log_norm = self.log_base_measure - self.log_partition
+        log_norm += sum(
+            dist.log_base_measure - dist.log_partition
+            for dist in dists)
+
+        # Calculate log normalisation of product of messages
+        prod_dist = self.sum_natural_parameters(*dists)
+        log_norm -= prod_dist.log_base_measure - prod_dist.log_partition
+
+        return log_norm
+
+    @staticmethod
+    def _iter_dists(dists):
+        for elem in dists:
+            if isinstance(elem, AbstractMessage):
+                yield elem
+            elif np.isscalar(elem):
+                yield elem
+            else:
+                for dist in elem:
+                    yield dist
+
+    def update_invalid(self, other: 'AbstractMessage') -> 'AbstractMessage':
+        valid = self.check_valid()
+        if self.ndim:
+            valid_parameters = (
+                np.where(valid, p, p_safe) for p, p_safe in zip(self, other))
+        else:
+            valid_parameters = self if valid else other  # TODO: Fairly certain this would not work
+        return type(self)(*valid_parameters, log_norm=self.log_norm)
+
+    def check_support(self) -> np.ndarray:
+        if self._parameter_support is not None:
+            return reduce(
+                and_,
+                ((p >= support[0]) & (p <= support[1])
+                 for p, support in zip(self.parameters, self._parameter_support)))
+        elif self.ndim:
+            return np.array(True, dtype=bool, ndmin=self.ndim)
+        return np.array([True])
+
+    def check_finite(self) -> np.ndarray:
+        return np.isfinite(self.natural_parameters).all(0)
+
+    def check_valid(self):
+        return self.check_finite() & self.check_support()
```

Comment:
Did you mean to use bitwise operators here and 332? Are the objects implementing some custom boolean logic (e.g. numpy)?

Reply:
yes, there's numpy boolean logic going on here